[Merged by Bors] - chore(data/rat): split `field ℚ` instance from definition of `ℚ`

Vierkantor opened the pull request:
I want to refer to the rational numbers in the definition of a field, meaning we can't define the field structure on `ℚ` in the same file as `ℚ` itself.

This PR moves everything in `data/rat/basic.lean` that does not depend on `algebra/field/basic.lean` into a new file `data/rat/defs.lean`: definition of the type `ℚ`, the operations giving its algebraic structure, and the coercions from integers and natural numbers. Basically, everything except the actual `field ℚ` instance.

It turns out our basic lemmas on rational numbers only require a `comm_ring`, `comm_group_with_zero` and `is_domain` instance, so I defined those instances in `defs.lean` could leave all lemmas intact.

As a consequence, the transitive imports provided by `data.rat.basic` are somewhat smaller: no `linear_ordered_field` is needed until `data.rat.order`. I see this as a bonus but can also re-import `algebra.order.field` in `data.rat.basic` if desired.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #14890 (but it's not a heavy dependency)
- [x] depends on: #14995 (because it redoes the exact same thing that #14890 did and got undone)
- [x] depends on: #14895 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
Fix import

Vierkantor committed:
Readjust imports

Vierkantor committed:
Add postponed import

Vierkantor committed:
Fix timeout in `number_theory/number_field.lean` (seems to be related to a non-defeq `algebra ℚ` instance?)

Vierkantor committed:
Add an explicit `rat.has_div` instance so expressions become literally equal

The test requires that the expression generated by `reflect` is literally equal
to a syntax quotation; due to having both a `comm_group_with_zero` instance
and a `field` instance there were now two defeq but not literally equal paths
being found. The easiest way to unify them is to ensure there is a high-priority
`has_div` instance.

Also adjusted the instances so all data-valued fields explicitly refer to the
high-priority instances, not their notation. I believe this can slightly help
defeq checks.

Vierkantor committed:
Help elaborator so we don't need `convert`

eric-wieser commented on src/data/rat/basic.lean:
```diff
@@ -0,0 +1,44 @@
+import logic.encodable.basic
+
+/-!
+# Basic lemmas for the Rational Numbers
```
I would claim that this file does not contain a single lemma!

eric-wieser commented on src/data/rat/defs.lean:
```diff
@@ -444,13 +445,17 @@ eq.trans (rat.mul_comm _ _) (rat.mul_inv_cancel _ h)
-instance : field ℚ :=
+/-! At this point in the import hierarchy we have not defined the `field` typeclass.
+Instead we'll instantiate `comm_ring` and `comm_group_with_zero` at this point.
+The `field` structure and any field-specific lemmas can be found in `data.rat.basic`.
```
```suggestion
The `rat.field` insstance and any field-specific lemmas can be found in `data.rat.basic`.
```
for the same of making a link in the docs

eric-wieser commented on src/data/rat/defs.lean:
```diff
@@ -461,17 +466,28 @@ instance : field ℚ :=
+{ .. rat.comm_group_with_zero,
+  .. (infer_instance : no_zero_divisors ℚ) }
 
 /- Extra instances to short-circuit type class resolution -/
```
These extra instances are no longer going to go via `rat.field`, which is probably why you're hitting elaboration problems elsewhere.

My hunch would be that everything would actually behave better if you assembled these shortcuts first, building stronger structures out of weaker ones; but it's entirely possible that would actually make things much worse.

Vierkantor replied:
There was only one real elaboration issue, in [counterexamples/pseudoelement.lean](https://github.com/leanprover-community/mathlib/pull/14893/files#diff-3c2f495ffe9b96910c43c53f664f803920d2135d04ea1df11d251694a5db9006), and that one used to have a painful elaboration workaround that is now much better. On the other hand, defining the weaker structures first means reproving results that are currently already deduced from others (see e.g. the `mul_zero` and `zero_mul` fields of the `add_comm_group_with_zero` instance). So I don't think it's worth it to redo all these definitions.

(Also I am not sure these "shortcut instances" are really worth it: see e.g. #1561)

eric-wieser commented on src/data/rat/defs.lean:
```diff
@@ -444,13 +445,17 @@ eq.trans (rat.mul_comm _ _) (rat.mul_inv_cancel _ h)
   one              := 1,
-  mul              := rat.mul,
-  inv              := rat.inv,
+  mul              := has_mul.mul,
```
```suggestion
  mul              := (*),
```

eric-wieser commented on src/data/rat/basic.lean:
```diff
@@ -0,0 +1,44 @@
+
+- `rat.field` is the field structure on `ℚ`.
+
+## Tags
```
This could probably do with an `## Implementation notes` section explaining why this needs to be in a separate file.

Vierkantor committed:
Split off everything not depending on the field structure on `ℚ`

Vierkantor committed:
Clarify notation

Vierkantor committed:
Make a link in the docstring

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Improve module docstring for `data.rat.basic`

Vierkantor committed:
chore(data/int/cast): redo #14890, moving field-specific lemmas

In #14894, I want to refer to the rational numbers in the definition of a field, meaning we can't have `algebra.field.basic` in the transitive imports of `data.rat.basic`.

Apparently this dependency was re-added, so I'm going to have to split it again...

Vierkantor committed:
Fix `number_theory/number_field.lean` (also `convert` the other half)

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into split-rat-defs
